hip/base/kernel_launch.hip.hpp

```diff
@@ -0,0 +1,261 @@
+struct matrix_accessor {
+    ValueType *data;
+    size_type stride;
+
+    __device__ ValueType &operator()(size_type row, size_type col)
+    {
+        return data[row * stride + col];
+    }
+
+    __device__ ValueType &operator[](size_type idx) { return data[idx]; }
+};
+
+
+template <typename T>
+struct device_map_impl {
+    using type = std::decay_t<hip_type<T>>;
+    static type map_to_device(T in) { return as_hip_type(in); }
+};
+
+template <typename ValueType>
+struct device_map_impl<matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<hip_type<ValueType>>;
+    static type map_to_device(matrix::Dense<ValueType> *mtx)
+    {
+        return {as_hip_type(mtx->get_values()), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct device_map_impl<const matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<const hip_type<ValueType>>;
+    static type map_to_device(const matrix::Dense<ValueType> *mtx)
+    {
+        return {as_hip_type(mtx->get_const_values()), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct device_map_impl<Array<ValueType> &> {
+    using type = hip_type<ValueType> *;
+    static type map_to_device(Array<ValueType> &array)
+    {
+        return as_hip_type(array.get_data());
+    }
+};
+
+template <typename ValueType>
+struct device_map_impl<const Array<ValueType> &> {
+    using type = const hip_type<ValueType> *;
+    static type map_to_device(const Array<ValueType> &array)
+    {
+        return as_hip_type(array.get_const_data());
+    }
+};
+
+
+template <typename T>
+typename device_map_impl<T>::type map_to_device(T &&param)
+{
+    return device_map_impl<T>::map_to_device(param);
+}
+
+
+template <typename ValueType>
+struct compact_dense_wrapper {
+    ValueType *data;
+};
+
+
+template <typename ValueType>
+compact_dense_wrapper<hip_type<ValueType>> compact(
+    matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {as_hip_type(mtx->get_values())};
+}
+
+
+template <typename ValueType>
+compact_dense_wrapper<const hip_type<ValueType>> compact(
+    const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {as_hip_type(mtx->get_const_values())};
+}
+
+
+template <typename ValueType>
+hip_type<ValueType> *vector(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return as_hip_type(mtx->get_values());
+}
+
+
+template <typename ValueType>
+const hip_type<ValueType> *vector(const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return as_hip_type(mtx->get_const_values());
+}
+
+
+template <typename T>
+struct device_unpack_1d_impl {
+    using type = T;
+    static __device__ __forceinline__ type unpack(T param, size_type, size_type)
```

Comment:
Can you elaborate? For the function attributes, the whole struct or the implementation? The former two should work, the latter would probably difficult, since this is used in a template context.